numpy/core/include/numpy/npy_cpu.h

```diff
@@ -109,11 +109,41 @@
     #error Unknown CPU, please report this to numpy maintainers with \
     information about your platform (OS, CPU and compiler)
 #endif
+/*
+Unaligned memory accesses occur when you try to read N bytes of data starting
+from an address that is not evenly divisible by N (i.e. addr % N != 0).
+For example, reading 4 bytes of data from address 0x10004 is fine, but
+reading 4 bytes of data from address 0x10005 would be an unaligned memory
+access.In reality, only a few architectures supports unaligned memory access.
+The effects of performing an unaligned memory access vary from architecture
+to architecture. It would be easy to write a whole document on the differences
+here; a summary of the common scenarios is presented below:
 
-#if (defined(NPY_CPU_X86) || defined(NPY_CPU_AMD64))
-#define NPY_CPU_HAVE_UNALIGNED_ACCESS 1
-#else
-#define NPY_CPU_HAVE_UNALIGNED_ACCESS 0
+ 1. Some architectures are able to perform unaligned memory accesses
+   transparently, but there is usually a significant performance cost.
+   (eg:X86, AMD64, ARM64, powerpc64)
+ 2. Some architectures raise processor exceptions when unaligned accesses
+   happen. The exception handler is able to correct the unaligned access,
+   at significant cost to performance.
+   (eg:ARM32, Alpha)
+ 3. Some architectures raise processor exceptions when unaligned accesses
+   happen, but the exceptions do not contain enough information for the
+   unaligned access to be corrected.
+   (eg:MIPS, Sparc)
+ 4. Some architectures are not capable of unaligned memory access, but will
+   silently perform a different memory access to the one that was requested,
+   resulting in a subtle code bug that is hard to detect!
+   (eg:RISCV, ia64)
+It should be obvious from the above that strong alignment should defined in
+all situations except No.1.
+
+NOTE: in x86 platform this flag can only be enabled if autovectorization is disabled.
+*/
+#if defined(NPY_CPU_X86) || defined(NPY_CPU_AMD64) || defined(__aarch64__) || defined(__powerpc64__)
```

Comment:
@Qiyu8, Its should be a simple comment not a case study, and the current one not accurate enough. the comment can be simple as following:
```C
// Except for the following architectures, memory access is limited to the natural
// alignment of data types otherwise it may lead to bus error or performance regression.
```
the new #def `NPY_STRONG_ALIGNMENT_REQUIRED` is too long, three words is enough `NPY_ALIGNMENT_REQUIRED` or `NPY_STRONG_ALIGNMENT`

Reply:
Sorry, I tried to shut down a bit over the break. I agree with Sayed, we don't need a list of how things can go wrong or how specific CPUs behave (it is interesting though!).

The more important thing is to write a short comment about where it is *safe* to do unaligned access (i.e. if you put a `if (!NPY_STRONG_ALIGNMENT_REQUIRED) {code}` what are you allowed to do in `code`? Are there remaining constraints about unaligned access?

Reply:
I will choose the name `NPY_ALIGNMENT_REQUIRED`,  which represents only aligned memory is allowed to access.

Reply:
Thanks. About that auto vectorization note in the comment? I do not have to be worried about it?

Reply:
`NPY_USE_UNALIGNED_ACCESS` already disabled no matter auto-vectorization is enabled or not, I will reconstruct [lowlevel_strided_loops.c.src](https://github.com/numpy/numpy/blob/master/numpy/core/src/multiarray/lowlevel_strided_loops.c.src) by using universal intrinsics so that `NPY_USE_UNALIGNED_ACCESS ` can enabled when -O3 is specifiled, but that's not the point of this PR.

Reply:
I realize that it is disabled, my question was just to make sure that the comment doesn't somehow apply to the new code as well, since I am not quite sure how it would differ from the one in `lowlevel_strided_loops.c.src`.

Btw. don't let it stop you, but that file will need at least two times bigger maintenance (one already open, the other is fixing most function signatures).  That should be pretty orthogonal to anything you do though, making sure casts are vectorized seems definitely worth it.

Reply:
Anyway, I am happy with this, feel free to merge. I just still don't quite understand the autovectorization comment unless it is specific to using `NPY_GCC_UNROLL_LOOPS`.